theRAPTLab/meme-2023
# Core platform update to Node version 18

This PR updates the core platform to Node JS 18 from Node JS 10 and also includes a required initial migration from Material UI v4 to v5.

**_Important_**: This work results in a stable system-level experience: the application will compile, run, serve content, and work in most respects. **The user interface has some aesthetic or functional regressions that are being addressed in follow-up work**.

## WHAT'S NEW
- NodeJS 18 has been introduced and several packages updated accordingly. For a complete list of system level package updates and considerations, see Issue #1.

## Testing Instructions
### Set-up
1. Pull the `dev-vertex/nodejs-18-upgrade` branch
2. Ensure you are using the correct node version. This project provides a `.nvmrc` file that specifies the recommended Node version, to use it (assuming you have _nvm_ installed): `nvm use`. It may ask require you to install the version prior to use - this can be done with `nvm install lts/hydrogen`
3. Do a first-pass installation of packages to support `npm run clean ...` scripts: `npm ci`
4. Run the clean: `npm run clean`. **Note**: this will remove any current data but retain the dataset templates.
5. Install the packages again: `npm ci`

You are now ready to test.

### Known Issues
- The Electron console does not render correctly. The application is still served and made available however.
- There are visual and other inconvenient defects that have been introduced by this update that relate to the migration of Material UI from Version 4 to Version 5. This will be addressed in follow-up work.

### Testing Instructions
Test the following scripts to confirm that they work as expected (compile, package, etc.) For the scripts that start/package the application, confirm the the application is available on the listed port and performs basic application functions.
 
- [ ] `npm run package`
     - [ ] Linux (should run, but Electron console does not render correctly. See "Known Issues")
- [ ] `npm run electron` (should run, but Electron console will not render correctly. See "Known Issues")
- [ ] `npm run start`
- [ ] `npm run clean`
- [ ] `npm run clean:all`
- [ ] `npm run appsign`
     - [ ] Linux (should run, but Electron console does not render correctly. See "Known Issues")
- [ ] `npm run app`

## Other Technical Notes
- "hydrogen" is the Long Term Support name for the NodeJS 18 version. This was chosen to allow patch/fixes to NodeJS to be applied. The LTS versions should not include breaking changes.
- Material UI version 5 adopted _Emotion_ as the default CSS rendering engine but allows for the usage of other engines. For the purposes of getting the application to a running system, we are currently using _styled components_ in lieu of JSS which was previously used.
